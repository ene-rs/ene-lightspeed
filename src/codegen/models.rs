use crate::models::entity::Entity;

use super::{contents::EntityTemplateContent, CodeGen};
use heck::{
    ToKebabCase, ToLowerCamelCase, ToShoutyKebabCase, ToShoutySnakeCase, ToSnakeCase, ToTitleCase,
    ToTrainCase, ToUpperCamelCase,
};
#[derive(Debug, Clone)]
pub struct EntityTemplate {
    pub entity_content: Vec<EntityTemplateContent>,
}

impl CodeGen for EntityTemplate {
    fn generate_code(&self, entity: &Entity, entities: &Vec<Entity>) -> anyhow::Result<String> {
        self.entity_content
            .iter()
            .map(|entity_content| entity_content.generate_code(entity, entities))
            .collect::<anyhow::Result<Vec<String>>>()
            .map(|entity_contents| entity_contents.join(""))
    }
}



#[derive(Debug, Clone)]
pub enum StringWithNamingConvention {
    KebabCase,
    LowerCamelCase,
    UpperCamelCase,
    ShoutyKebabCase,
    ShoutySnakeCase,
    SnakeCase,
    TitleCase,
    TrainCase,
}

impl StringWithNamingConvention {
    pub fn to_naming_convention(&self, value: &str) -> String {
        match self {
            Self::KebabCase => value.to_kebab_case(),
            Self::LowerCamelCase => value.to_lower_camel_case(),
            Self::UpperCamelCase => value.to_upper_camel_case(),
            Self::ShoutyKebabCase => value.to_shouty_kebab_case(),
            Self::ShoutySnakeCase => value.to_shouty_snake_case(),
            Self::SnakeCase => value.to_snake_case(),
            Self::TitleCase => value.to_title_case(),
            Self::TrainCase => value.to_train_case(),
        }
    }
}


// region: tests
#[cfg(test)]
mod tests {
    use crate::{codegen::CodeGen, models::entity::{AttributeType, Entity, ForeignKey, Semantics}, parsers::template_parser::TemplateParser};

    fn prepare_entities() -> Vec<Entity> {
         let car = Entity {
            name: "Car".to_string(),
            semantics: Semantics {
                plural: "Cars".to_string(),
            },
            attributes: vec![
                ("id".to_string(), AttributeType::Type("Uuid".to_string())),
                (
                    "name".to_string(),
                    AttributeType::Type("String".to_string()),
                ),
                (
                    "model".to_string(),
                    AttributeType::Type("String".to_string()),
                ),
                (
                    "manufacturer".to_string(),
                    AttributeType::Type("String".to_string()),
                ),
                ("year".to_string(), AttributeType::Type("i32".to_string())),
                (
                    "owner".to_string(),
                    AttributeType::ForeignKey(ForeignKey {
                        entity_name: "Person".to_string(),
                        attribute_name: "id".to_string(),
                    }),
                ),
            ]
            .into_iter()
            .collect(),
            primary_key: "id".to_string(),
            filter_by: vec![
                vec!["id".to_string()],
                vec!["name".to_string(), "model".to_string()],
                vec!["manufacturer".to_string()],
                vec!["year".to_string()],
                vec!["owner".to_string()],
            ],
            unique_attributes: vec![vec!["id".to_string()], vec!["name".to_string()]]
                .into_iter()
                .collect(),
        };
        let person = Entity {
            name: "Person".to_string(),
            semantics: Semantics {
                plural: "People".to_string(),
            },
            attributes: vec![
                ("id".to_string(), AttributeType::Type("Uuid".to_string())),
                (
                    "name".to_string(),
                    AttributeType::Type("String".to_string()),
                ),
                ("age".to_string(), AttributeType::Type("i32".to_string())),
            ]
            .into_iter()
            .collect(),
            primary_key: "id".to_string(),
            filter_by: vec![
                vec!["id".to_string()],
                vec!["name".to_string()],
                vec!["age".to_string()],
            ],
            unique_attributes: vec![vec!["id".to_string()], vec!["name".to_string()]]
                .into_iter()
                .collect(),
        };
        let entities = vec![car.clone(), person];
        entities
    }

    #[test]
    fn test_entity_code_gen() {
        let entity_model_test = std::fs::read_to_string("/Users/abdullahsabaaallil/Desktop/ene-lightspeed/template-rs/src/models/entity_name.rs").unwrap();
        let parsed_template = TemplateParser::parse_template(&entity_model_test).unwrap();
        let entities = prepare_entities();
        let generated_code = parsed_template.generate_code(&entities[0], &entities).unwrap();
        std::fs::write("test_output/entity_test.rs", generated_code.clone())
            .map_err(|e| {
                println!("Error: {:?}", e);
                e
            })
            .unwrap();
        insta::assert_debug_snapshot!(generated_code);
    }

    #[test]
    fn test_controller_code_gen() {
        let controller_test = std::fs::read_to_string("/Users/abdullahsabaaallil/Desktop/ene-lightspeed/template-rs/src/controllers/entity_controller.rs").unwrap();
        let parsed_template = TemplateParser::parse_template(&controller_test).unwrap();
        let entities = prepare_entities();
        let generated_code = parsed_template.generate_code(&entities[0], &entities).unwrap();
        std::fs::write("test_output/entity_controller_test.rs", generated_code.clone())
            .map_err(|e| {
                println!("Error: {:?}", e);
                e
            })
            .unwrap();
        insta::assert_debug_snapshot!(generated_code);
    }


}