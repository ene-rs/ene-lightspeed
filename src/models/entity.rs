use std::{collections::BTreeMap, fmt::Display};

use anyhow::bail;
use serde::{Deserialize, Serialize};

pub type AttributeName = String;
pub type FilterByAttributeNames = Vec<AttributeName>;

pub type FilterByAttributes = Vec<(AttributeName, AttributeType)>;

pub trait MostSpecificFilterAttributes {
    fn get_most_specific_filter_attribute(&self) -> (AttributeName, AttributeType);
}

impl MostSpecificFilterAttributes for &FilterByAttributes {
    fn get_most_specific_filter_attribute(&self) -> (AttributeName, AttributeType) {
        self.last().unwrap().clone()
    }
}
pub type UniqueAttributes = Vec<AttributeName>;

#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(untagged)]
pub enum AttributeType {
    Type(String),
    ForeignKey(ForeignKey),
}

impl AttributeType {
    pub fn get_attribute_type(&self, entities: &Vec<Entity>) -> anyhow::Result<String> {
        match self {
            AttributeType::Type(type_name) => Ok(type_name.clone()),
            AttributeType::ForeignKey(ForeignKey {
                entity_name,
                attribute_name,
            }) => {
                let foreign_key_entity = entities
                    .iter()
                    .find(|entity| entity.name == *entity_name)
                    .expect(&format!(
                        "Foreign key entity {} is not present in the entities",
                        entity_name
                    ));
                let foreign_key_attribute: &AttributeType = foreign_key_entity
                    .attributes
                    .get(attribute_name)
                    .expect(&format!(
                        "Foreign key attribute {} is not present in the attributes of {}",
                        attribute_name, entity_name
                    ));
                foreign_key_attribute.get_attribute_type(entities)
            }
        }
    }
}
#[derive(Serialize, Deserialize, Clone, Default)]
pub struct Entity {
    pub name: String,
    pub attributes: BTreeMap<AttributeName, AttributeType>,
    pub primary_key: AttributeName,
    pub filter_by: Vec<FilterByAttributeNames>,
    pub unique_attributes: Vec<UniqueAttributes>,
    pub semantics: Semantics,
}

impl Entity {
    /**
     * Possible constraints:
     * - All attributes used as primary key, filter by, unique attributes, or foreign keys must be present
     * - Primary key must be in the attributes
     * - If there are unique attributes, they need to be present in filter_by
     * - If an attribute is unique, it cannot be used as a sub attribute in filter_by
     * For Example:
     * name is a unique attribute
     * It is not possible to have a filter_by with: [id, name], since name is already unique
     */
    pub fn validate_request(&self, entities: Vec<Entity>) -> anyhow::Result<()> {
        let attribute_names = self.attributes.keys().collect::<Vec<&String>>();
        if !attribute_names.contains(&&self.primary_key) {
            bail!(
                "Primary key {} is not present in the attributes of {}",
                self.primary_key,
                self.name
            );
        }

        for foreign_key in self.get_foreign_keys().iter() {
            // search the foreign key in the other entities
            let foreign_key_entity = entities
                .iter()
                .find(|entity| entity.name == foreign_key.entity_name)
                .expect(&format!(
                    "Foreign key entity {} is not present in the entities",
                    foreign_key.entity_name
                ));
            if !foreign_key_entity
                .attributes
                .iter()
                .map(|(attribute_name, _)| attribute_name)
                .collect::<Vec<&String>>()
                .contains(&&foreign_key.attribute_name)
            {
                bail!(
                    "Foreign key attribute {} is not present in the attributes of {}",
                    foreign_key.attribute_name,
                    foreign_key.entity_name
                );
            }
        }

        for unique_attribute in self
            .unique_attributes
            .iter()
            .flatten()
            .collect::<Vec<&String>>()
        {
            if self.attributes.get(unique_attribute).is_none() {
                bail!(
                    "Unique attribute {} is not present in the attributes of {}",
                    unique_attribute,
                    self.name
                );
            }
        }
        for filter_by_attribute in self.filter_by.iter().flatten().collect::<Vec<&String>>() {
            if self.attributes.get(filter_by_attribute).is_none() {
                bail!(
                    "Filter by attribute {} is not present in the attributes of {}",
                    filter_by_attribute,
                    self.name
                );
            }
        }

        // Verify that the unique attributes are present in the filter_by
        for unique_attribute in self.unique_attributes.iter() {
            if !self.filter_by.contains(unique_attribute) {
                bail!(
                    "Unique attribute {:?} is not present in the filter_by of {}",
                    unique_attribute,
                    self.name
                );
            }

            let common = self
                .filter_by
                .iter()
                .filter(|filter_by| filter_by != &unique_attribute)
                .any(|filter_by| {
                    println!("{:?}", filter_by);
                    is_sub(&filter_by, &unique_attribute)
                });

            if common {
                bail!("Unique attribute {:?} is a sub attribute of another filter_by in {}. It does not make sense to have fine grained filters on unique attributes, since they're unique", unique_attribute, self.name);
            }
        }

        Ok(())
    }

    pub fn get_foreign_keys(&self) -> Vec<ForeignKey> {
        let mut foreign_keys = Vec::new();
        for (_, attribute_type) in &self.attributes {
            match attribute_type {
                AttributeType::ForeignKey(foreign_key) => {
                    foreign_keys.push(foreign_key.clone());
                }
                _ => {}
            }
        }
        foreign_keys
    }

    pub fn get_filter_by_attributes(&self) -> Vec<FilterByAttributes> {
        let mut filter_by_attributes = Vec::new();
        for filter_by in &self.filter_by {
            let mut filter_by_attribute = Vec::new();
            for attribute_name in filter_by {
                filter_by_attribute.push((
                    attribute_name.clone(),
                    self.attributes.get(attribute_name).unwrap().clone(),
                ));
            }
            filter_by_attributes.push(filter_by_attribute);
        }
        filter_by_attributes
    }
}

fn is_sub<T: PartialEq>(mut haystack: &[T], needle: &[T]) -> bool {
    if needle.len() == 0 {
        return true;
    }
    while !haystack.is_empty() {
        if haystack.starts_with(needle) {
            return true;
        }
        haystack = &haystack[1..];
    }
    false
}

#[derive(Serialize, Deserialize, Clone, Default)]
pub struct Semantics {
    pub plural: String,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ForeignKey {
    pub entity_name: String,
    pub attribute_name: String,
}

impl Display for AttributeType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}
