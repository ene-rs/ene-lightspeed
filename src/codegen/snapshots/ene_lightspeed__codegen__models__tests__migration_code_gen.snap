---
source: src/codegen/models.rs
expression: generated_code
---
"\nCREATE TABLE IF NOT EXISTS car (\n    id uuid,\n    manufacturer string,\n    model string,\n    name string,\n    owner uuid,\n    year i32\n)\nCREATE INDEX IF NOT EXISTS unique_id_index ON (\n    id\n)\nCREATE INDEX IF NOT EXISTS unique_model_index ON (\n    name,\n    model\n)\nCREATE INDEX IF NOT EXISTS unique_manufacturer_index ON (\n    manufacturer\n)\nCREATE INDEX IF NOT EXISTS unique_year_index ON (\n    year\n)\nCREATE INDEX IF NOT EXISTS unique_owner_index ON (\n    owner\n)\nCREATE UNIQUE INDEX IF NOT EXISTS unique_id_index ON (\n    id\n)\nCREATE UNIQUE INDEX IF NOT EXISTS unique_name_index ON (\n    name\n)"
